File: pages/data-structure/linear-structure/code/stack.py
class Stack:
    def __init__(self, max_size=100):
        self.data = [None] * max_size  # 顺序存储的栈
        self.top = -1                  # 栈顶指针
        self.max_size = max_size       # 栈最大容量

    def is_empty(self):
        return self.top == -1

    def push(self, value):
        if self.top >= self.max_size - 1:
            raise OverflowError("栈已满")
        self.top += 1
        self.data[self.top] = value

    def pop(self):
        if self.is_empty():
            raise IndexError("Stack underflow")
        value = self.data[self.top]
        self.top -= 1
        return value

    def peek(self):
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self.data[self.top]

    def print_stack(self):
        print(self.data[:self.top + 1])

File: pages/data-structure/linear-structure/code/stack.c
#include <stdio.h>
#include <stdbool.h>
#define MAX_SIZE 100

typedef struct {
    int data[MAX_SIZE];  // 数组存储栈元素
    int top;             // 栈顶指针
} Stack;

void InitStack(Stack* s) {
    s->top = -1;
}

bool IsEmpty(Stack* s) {
    return s->top == -1;
}

bool Push(Stack* s, int value) {
    if (s->top >= MAX_SIZE - 1) return false;
    s->data[++s->top] = value;
    return true;
}

bool Pop(Stack* s, int* value) {
    if (IsEmpty(s)) return false;
    *value = s->data[s->top--];
    return true;
}

bool Peek(Stack* s, int* value) {
    if (IsEmpty(s)) return false;
    *value = s->data[s->top];
    return true;
}

void PrintStack(Stack* s) {
    for (int i = 0; i <= s->top; i++) {
        printf("%d ", s->data[i]);
    }
    printf("\n");
}
